### PR TITLE
Upgrade Jetty 9.4.x to Junit 5

### PR DESCRIPTION
**Pull request by joakime**

### CODE REVIEW

File: jetty-alpn/jetty-alpn-java-server/src/test/java/org/eclipse/jetty/alpn/java/server/JDK9ALPNTest.java [hunk 1]
```diff
@@ -109,7 +111,7 @@ protected void doNonErrorHandle(String target, Request baseRequest, HttpServletR
             InputStream input = client.getInputStream();
             BufferedReader reader = new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8));
             String line = reader.readLine();
-            Assert.assertTrue(line.contains(" 200 "));
+            assertTrue(line.contains(" 200 "));
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-alpn/jetty-alpn-java-server/src/test/java/org/eclipse/jetty/alpn/java/server/JDK9ALPNTest.java]: Nah.  The new jupiter Assertions class does a much better job (then Junit 4's Assert class) of reporting the details of a test failure.

https://junit.org/junit5/docs/current/api/org/junit/jupiter/api/Assertions.html

assertThat() is still available, as it comes from the hamcrest side of junit.
If you feel more comfortable writing in that style, go for it.
(I even suspect that the opentest4j features are available too in junit5)

The biggest change between Assert and Assertions is the order of parameters.

eg:

junit 4 | junit 5
------ | -------
assertEquals(message, expected, actual) | assertEquals(expected, actual, message)
assertTrue(message, boolean) | assertEquals(boolean, message)
assertNull(message, Object) | assertEquals(Object, message)

It's easy to have an old `assertEquals("These two should be the same", expectedStr, actualStr)` quietly/mysteriously fail as the `message` portion is at the end of the parameter list now.
Thankfully the migration/refactoring tools took care of the majority of these.

[the remark this answers is not in the document]
Comment (joakime) [reply jetty-alpn/jetty-alpn-java-server/src/test/java/org/eclipse/jetty/alpn/java/server/JDK9ALPNTest.java]: As for that line, you could change it to ...

``` java
assertThat(line, containsString(" 200 "));
```

Which, when it fails, will report the entirety of value of `line` and what was being looked for, as well as the line #### it occurred on (in the test)

But if you used ...

``` java
assertTrue(line.contains(" 200 "));
```

Then the failure will only mention that an assertion on true failed on line ####

[the remark this answers is not in the document]
Comment (gregw) [reply jetty-alpn/jetty-alpn-java-server/src/test/java/org/eclipse/jetty/alpn/java/server/JDK9ALPNTest.java]: I think the reporting of failed values is important, so we should take the opportunity to change `assertTrue(someString.contains(subString))` to `assertThat(someString,containsString(subString))`

**Comment by joakime** [comment]
I'll look at updating this and merging it back into `jetty-9.4.x` when I get back from vacation. (May 21st)

**Comment by joakime** [comment]
@olamy I think we should ban vintage junit jars/apis from everything (except the OSGi tests, due to paxexam)

File: pom.xml [hunk 2]
```diff
@@ -29,7 +29,7 @@
     <asm.version>6.2</asm.version>
     <jmh.version>1.20</jmh.version>
     <jmhjar.name>benchmarks</jmhjar.name>
-    <surefire.version>3.0.0-SNAPSHOT</surefire.version>
+    <surefire.version>2.22.0</surefire.version>
```
Comment (joakime) [inline pom.xml]: oooh. 👍

Comment (joakime) [reply pom.xml]: Does this work with jacoco?
Does this work with -T3 (currently disabled in this branch)?
Does this work with `@{argLine}` (yes, that's the correct syntax per the documentation and intellij support.)?

Comment (olamy) [reply pom.xml]: jacoco: sure just have a look at the ci :-) https://jenkins.webtide.net/job/jetty.project/view/change-requests/job/PR-2436/
-T3 I can't see why it couldn't work. do you have any special log/error indicating not? (restored that and CI seems happy)
@{argLine} yup it works sorry I made test with a bad surefire snapshot version (restored and CI seems happy)

**Comment by joakime** [comment]
This is planned to be merged post 9.4.12 release.

File: Jenkinsfile [hunk 3]
```diff
@@ -98,7 +98,7 @@ def getFullBuild(jdk, os) {
                       //options: [invokerPublisher(disabled: false)],
                       mavenOpts: mavenOpts,
                       mavenLocalRepo: localRepo) {
-                sh "mvn -V -B install -Dmaven.test.failure.ignore=true -e -Pmongodb -T3 -DmavenHome=${mvntoolInvoker} -Dunix.socket.tmp="+env.JENKINS_HOME
+                sh "mvn -U -V -B install -Denv=ci -Dmaven.test.failure.ignore=true -e -Pmongodb -T3 -DmavenHome=${mvntoolInvoker} -Dunix.socket.tmp="+env.JENKINS_HOME
```
Comment (joakime) [inline Jenkinsfile]: Do we still need the `-U` option here?

Comment (olamy) [reply Jenkinsfile]: nope not using anymore surefire snapshot

**Pull request merged**